Bryce/azure advanced configuration (#17910)
DataDog/documentation: content/en/integrations/guide/azure-architecture-and-configuration.md

---
title: Azure Architecture and Configuration
kind: guide
description: "Guide to architecture of the Datadog Azure integration and alternate configuration options"
further_reading:
- link: "https://docs.datadoghq.com/integrations/azure/"
  tag: "Documentation"
  text: "Azure Integration"
---

## Overview

> This guide provides detailed information and reference architectures for users configuring Datadog’s Azure integration, as well as alternative configuration options for specific use cases.

Reviewer: ```suggestion
This guide provides detailed information and reference architectures for users configuring Datadog's Azure integration, as well as alternative configuration options for specific use cases.
```

> ### Reference Architectures

Reviewer: ```suggestion
### Reference architectures
```

> The diagrams in this guide provide a visual representation of the configuration process and outcome when following the steps in the [Azure Integration page][1]. This provides a more detailed overview of Datadog's interaction with your Azure environment, and answers common security, compliance, and governance questions.

Reviewer: ```suggestion
The diagrams in this guide provide a visual representation of the configuration process and outcome when following the steps in the [Azure integration page][1]. This guide provides a detailed overview of Datadog's interaction with your Azure environment and answers common security, compliance, and governance questions.
```

> ### Alternate Configurations

Reviewer: ```suggestion
### Alternate configurations
```

> The setup processes documented in the [Azure Integration page][1] are the recommended steps and result in the ideal configuration for the majority of users. Alternate configuration options in this document may be preferable for certain use cases, and any trade-offs in performance, features, or ease-of-management are outlined as needed.

Reviewer: ```suggestion
The setup processes documented in the [Azure integration page][1] are the recommended steps and result in the ideal configuration for the majority of users. Alternate configuration options in this document may be preferable for certain use cases. Any trade-offs in performance, features, or ease-of-management are outlined as needed.
```

> ## Azure Metric and Data Collection

Reviewer: ```suggestion
## Azure metric and data collection
```

> Enabling Datadog’s Azure integration allows Datadog to:

Reviewer: ```suggestion
Enabling Datadog's Azure integration allows Datadog to:
```

  - Discover and monitor all resources in all subscriptions within the given scope
  - Automatically update the metric definitions, to ensure that all of the metrics available from Azure Monitor are collected
  - Ingest a range of both general and resource-specific metadata (including custom Azure tags), and apply it to the associated resource metrics in Datadog as tags

>   - Query Azure Metadata APIs and use the responses to [generate useful metrics in Datadog][2] for insights that Azure Monitor does not support

Reviewer: ```suggestion
  - Query Azure metadata APIs and use the responses to [generate useful metrics in Datadog][2] for insights that Azure Monitor does not support
```

> The Azure APIs used and data collected are identical regardless of whether you use the standard or Azure Native version of the integration.

Reviewer: ```suggestion
The Azure APIs used and data collected are identical regardless of whether you use the standard or Azure native version of the integration.
```

{{% site-region region="us,us5,eu,gov,ap1" %}}

> ### Standard Azure Integration Metric and Data Collection

Reviewer: ```suggestion
### Standard Azure integration metric and data collection
```

_Available in all Datadog Sites_

Follow these steps to enable the standard Azure integration:

  1. Create an app registration in your Active Directory and enter the credentials in the [Datadog Azure integration page][2].

>   2. Give the application read-access (`Monitoring Reader` role) to the subscriptions or management group you would like to monitor.

Reviewer: ```suggestion
  2. Give the application read access (`Monitoring Reader` role) to the subscriptions or management group you would like to monitor.
```

> The diagram below outlines the process and resulting architecture of the Azure Integration configuration described in the main documentation.

Reviewer: ```suggestion
The diagram below outlines the process and resulting architecture of the Azure integration configuration described in the main documentation.
```

> {{< img src="integrations/guide/azure_architecture_and_configuration/app_registration_integration_setup.png" alt="Diagram of the App Registration integration setup" >}}

Reviewer: ```suggestion
{{< img src="integrations/guide/azure_architecture_and_configuration/app_registration_integration_setup.png" alt="Diagram of the app registration integration setup" >}}
```

> Once this is completed, data collection begins automatically. The App Registration information entered in Datadog allows Datadog to [request a token from Azure Active Directory][1] (AD). Datadog uses this token as the authorization for API calls to various Azure APIs, to discover resources within the scope provided and collect data. This continuous process runs with two-minute intervals by default, and is used to discover and collect data from your Azure environment. This process is pictured below.

Reviewer: ```suggestion
Once this is completed, data collection begins automatically. The app registration information entered in Datadog allows Datadog to [request a token from Azure Active Directory][1] (AD). Datadog uses this token as the authorization for API calls to various Azure APIs, to discover resources within the scope provided, and collect data. This continuous process runs with two-minute intervals by default, and is used to discover and collect data from your Azure environment. The data collection process is pictured below.
```

{{< img src="integrations/guide/azure_architecture_and_configuration/app_registration_metric_collection.png" alt="Diagram of the App Registration integration setup" >}}

[1]: https://learn.microsoft.com/en-us/azure/databricks/dev-tools/api/latest/aad/
[2]: https://app.datadoghq.com/integrations/azure
{{% /site-region %}}
{{% site-region region="us3" %}}

> ### Azure Native integration metric collection

Reviewer: ```suggestion
### Azure native integration metric collection
```

_Available only in the Datadog US3 site_

> Follow these steps to enable the Azure Native integration:

Reviewer: ```suggestion
Follow these steps to enable the Azure native integration:
```

1. Confirm that your Datadog organization is hosted on the US3 [Datadog site][1], or [create a trial Datadog account on the US3 site][5].
2. Create a Datadog resource in Azure that links at least one subscription.
3. Optionally, update the Datadog resource to include other subscriptions.

> The Datadog resource in Azure is a link between your Azure environment and your Datadog account. This link enables the same data collection as the standard Azure integration available for other Datadog sites, but with a different authentication mechanism. Its access is assigned using a **System Managed Identity** associated with the Datadog resource in Azure, rather than a user-created and configured **App Registration**.

Reviewer: ```suggestion
The Datadog resource in Azure links your Azure environment and your Datadog account. This link enables the same data collection as the standard Azure integration available for other Datadog sites, but with a different authentication mechanism. Its access is assigned using a **System Managed Identity** associated with the Datadog resource in Azure, rather than a user-created and configured **App Registration**.
```

> When the Datadog resource is created, its resource ID and other information is passed to Datadog. Datadog uses its own App Registration to request a token from Azure AD, authenticating with Azure as Datadog. Datadog then uses this Datadog token along with the user's Datadog resource ID to request a user token from a bespoke Azure service created for the Azure Native integration. This Azure service verifies Datadog’s identity, confirms the user has granted access with the Datadog resource, and then returns a short-lived token to Datadog. Datadog uses this token as the authorization for API calls made to Azure APIs in the user's environment, to discover resources within the scope provided and collect data.

Reviewer: ```suggestion
After you create the Datadog resource, the following steps occur:
1. The Azure environment passes the Datadog resource's resource ID and other information to Datadog.
1. Datadog uses its own App Registration to request a token from Azure AD, authenticating with Azure as Datadog. 
1. Datadog uses the Datadog token along with the user's Datadog resource ID to request a user token from a bespoke Azure service created for the Azure Native integration.
1. The Azure service verifies Datadog's identity, confirms the user has granted access with the Datadog resource, and then returns a short-lived token to Datadog.
1. Datadog uses the token to authorize API calls made to Azure APIs in the user's environment. Datadog uses API calls to discover resources within the scope provided and collect data.
```

The `Monitoring Reader` role assignment happens automatically during the creation of the Datadog resource, and is scoped to the parent subscription of the Datadog resource. If you add additional subscriptions for monitoring to the Datadog resource, this scope is updated for the Managed Identity automatically.

> The diagram below outlines the process and resulting architecture of the Azure Native Integration configuration.

Reviewer: ```suggestion
The diagram below outlines the process and resulting architecture of the Azure native integration configuration.
```

> {{< img src="integrations/guide/azure_architecture_and_configuration/azure_native_integration_setup.png" alt="Diagram of the Azure Native integration setup" >}}

Reviewer: ```suggestion
{{< img src="integrations/guide/azure_architecture_and_configuration/azure_native_integration_setup.png" alt="Diagram of the Azure native integration setup" >}}
```

Once this is completed, data collection begins automatically. Datadog continuously discovers and collects metrics from your Azure environment, as pictured below.

> {{< img src="integrations/guide/azure_architecture_and_configuration/azure_native_metric_collection.png" alt="Diagram of the Azure Native metric collection setup" >}}

Reviewer: ```suggestion
{{< img src="integrations/guide/azure_architecture_and_configuration/azure_native_metric_collection.png" alt="Diagram of the Azure native metric collection setup" >}}
```

### Alternate Configuration Options for Log Forwarding with the Azure Native Integration

The one-click buttons to enable logs in the Datadog resource automate the process of adding diagnostic settings. In some cases, organizations may want to manage and configure diagnostic settings themselves, while still taking advantage of the automated log forwarding capability with the Azure Native integration. 

Manually created diagnostic settings are not impacted by log settings on the Datadog resource, and are not deleted based on the tag rules specified in the Datadog resource. Resource logs do not have to be enabled on the Datadog resource in order for manual log forwarding to work. However, the Datadog resource being used for log forwarding must not be in a disabled state.

Reasons for manually managing Diagnostic Settings include:

  1. Infrastructure-as-code Policies
       Strict internal policies around IaC that require all resources to to be created and managed deterministically (for example, if automatic creation of diagnostic settings by the Datadog resource would cause an unresolvable conflict between the desired and the actual state).

>   2. Limiting Resource Log Categories

Reviewer: ```suggestion
  2. Limiting resource log categories
```

       As the diagnostic settings created automatically by the Datadog resource include all log categories, specifying a subset of these categories requires you to create diagnostic settings yourself.  
       **Note**: You can also use [exclusion filters][1] to exclude these logs from being indexed upon ingestion into Datadog.

  3. Cross-subscription log forwarding

>        This can be useful if you only want to send logs and no other data from a specific subscription. The Datadog resource used for log forwarding still collects metrics and data from its own subscription. You must register the Resource Provider `Microsoft.Datadog` in each subscription intended to send logs before creating the diagnostic settings.

Reviewer: ```suggestion
       This can be useful if you only want to send logs and no other data from a specific subscription. The Datadog resource used for log forwarding still collects metrics and data from its own subscription. You must register the resource provider `Microsoft.Datadog` in each subscription intended to send logs before creating the diagnostic settings.
```

>        {{< img src="integrations/guide/azure_architecture_and_configuration/datadog_agent_build_resource_providers.png" alt="Screenshot of the Resource providers page in the Azure Portal with Microsoft.Datadog showing the status of registered" >}}

Reviewer: ```suggestion
       {{< img src="integrations/guide/azure_architecture_and_configuration/datadog_agent_build_resource_providers.png" alt="Screenshot of the resource providers page in the Azure Portal with Microsoft. Datadog showing the status of registered." >}}
```

  4. Testing 
       Sending example logs to Datadog can be useful for testing or other investigations. In these cases, adding diagnostic settings manually can be quicker than waiting for them to be created automatically from updated tags and settings.

This architecture is shown below, including the optional cross-subscription set up:

> {{< img src="integrations/guide/azure_architecture_and_configuration/custom_azure_native_log_forwarding.png" alt="Diagram of the custom Azure Native log forwarding setup" >}}

Reviewer: ```suggestion
{{< img src="integrations/guide/azure_architecture_and_configuration/custom_azure_native_log_forwarding.png" alt="Diagram of the custom Azure native log forwarding setup" >}}
```

---

> title: Azure Architecture and Configuration

Reviewer 2: Azure Integration Architecture and Configuration

  tag: "Documentation"
  text: "Azure Integration"
---

## Overview

This guide provides detailed information and reference architectures for users configuring Datadog's Azure integration, as well as alternative configuration options for specific use cases.

### Reference architectures

The diagrams in this guide provide a visual representation of the configuration process and outcome when following the steps in the [Azure integration page][1]. This guide provides a detailed overview of Datadog's interaction with your Azure environment and answers common security, compliance, and governance questions.

### Alternate configurations

The setup processes documented in the [Azure integration page][1] are the recommended steps and result in the ideal configuration for the majority of users. Alternate configuration options in this document may be preferable for certain use cases. Any trade-offs in performance, features, or ease-of-management are outlined as needed.

## Azure metric and data collection

Enabling Datadog's Azure integration allows Datadog to:

  - Discover and monitor all resources in all subscriptions within the given scope

>   - Automatically update the metric definitions, to ensure that all of the metrics available from Azure Monitor are collected

Reviewer 2: Remove "the" in front of metric definitions, or maybe rewrite so it's "Automatically update discovered metric definitions"

**Permissions**: The `Monitoring Reader` role assignment happens automatically during the creation of the Datadog resource, and is scoped to the parent subscription of the Datadog resource. If you add additional subscriptions for monitoring to the Datadog resource, this scope is updated for the Managed Identity automatically.

Follow these steps to enable the Azure Native integration:

1. Confirm that your Datadog organization is hosted on the US3 [Datadog site][1], or [create a trial Datadog account on the US3 site][5].
2. Create a Datadog resource in Azure that links at least one subscription.
3. Optionally, update the Datadog resource to include other subscriptions.

As an [external ISV][6], there is an additional, separate process to request and use this access:

1. Datadog authenticates into Azure and uses a private Azure service to request the customer token associated with the given Datadog resource.

> 1. This Azure service verifies Datadog’s identity, and ensures the requested Datadog resource exists and is enabled.

Reviewer: ```suggestion
1. This Azure service verifies Datadog's identity, and ensures the requested Datadog resource exists and is enabled.
```
